mlflow/types/utils.py

```diff
@@ -17,45 +17,60 @@ def __init__(self, msg):
         )
 
 
+def _get_tensor_shape(data: np.ndarray) -> tuple:
+    """
+    Infer the shape of the inputted data.
+
+    This method creates the shape of the tensor to store in the TensorSpec. Always returns a shape
+    including a variable dimension. Assumes the variable dimension is the first dimension.
+
+    NB: For 1D tensors, the output is (-1, shape).
+
+    :param data: Dataset to infer from.
+    :return: tuple : Shape of the inputted data (including a variable dimension)
+    """
+    if not isinstance(data, np.ndarray):
+        raise TypeError("Expected numpy.ndarray, got '{}'.".format(type(data)))
+    if len(data.shape) == 1:
+        variable_input_data_shape = [-1, data.shape[0]]
+    else:
+        variable_input_data_shape = list(data.shape)
+        variable_input_data_shape[0] = -1
+    return tuple(variable_input_data_shape)
+
+
 def _infer_schema(data: Any) -> Schema:
     """
     Infer an MLflow schema from a dataset.
 
-    This method captures the column names and data types from the user data. The signature
-    represents model input and output as data frames with (optionally) named columns and data
-    type specified as one of types defined in :py:class:`DataType`. This method will raise
-    an exception if the user data contains incompatible types or is not passed in one of the
-    supported formats (containers).
+    Data inputted as a numpy array or a dictionary is represented by :py:class:`TensorSpec`.
+    All other inputted data types are specified by :py:class:`ColSpec`.
+
+    A `TensorSpec` captures the data shape (default variable axis is 0), the data type (numpy.dtype)
+    and an optional name to represent the dataset.
+    A `ColSpec` captures the data type (defined in :py:class:`DataType`) and an optional name for
+    each individual column of the dataset.
+
+    This method will raise an exception if the user data contains incompatible types or is not
+    passed in one of the supported formats (containers).
 
     The input should be one of these:
       - pandas.DataFrame or pandas.Series
       - dictionary of { name -> numpy.ndarray}
       - numpy.ndarray
       - pyspark.sql.DataFrame
 
-    The element types should be mappable to one of :py:class:`mlflow.models.signature.DataType`.
```

Comment:
Lets leave this message in, just say mappable to one of the DataType for dataframes and one of the numpy types for tensors.